File: codes/semana11/exercicio02.cpp
// ✅ Exercício 2 — Configuração Singleton

#include <iostream>
#include <string>
#include <map>

// ── Singleton ────────────────────────────────────────────
class Configuracao {
    std::map<std::string, std::string> dados_ = {
        {"dificuldade", "normal"},
        {"volume",      "80"},
        {"idioma",      "pt-BR"}
    };
    Configuracao() = default;

public:
    Configuracao(const Configuracao&)            = delete;
    Configuracao& operator=(const Configuracao&) = delete;

    static Configuracao& instancia() {
        static Configuracao inst;
        return inst;
    }

    void alterar(const std::string& chave,
                 const std::string& valor) {
        dados_[chave] = valor;
    }

    std::string obter(const std::string& chave) const {
        return dados_.at(chave);
    }

    void exibir() const {
        for (const auto& [k, v] : dados_)
            std::cout << "  [Config] "
                      << k << " = " << v << "\n";
    }
};

// ── Classes consumidoras ─────────────────────────────────
class MenuPrincipal {
public:
    void aplicar_configuracoes() {
        auto& cfg = Configuracao::instancia();
        cfg.alterar("dificuldade", "dificil");
        cfg.alterar("volume",      "60");
        std::cout << "[Menu] Configurações alteradas"
                     " pelo jogador\n";
        cfg.exibir();
    }
};

class Masmorra {
public:
    void iniciar() {
        std::cout << "\n[Masmorra] Lendo configurações"
                     " do jogo...\n";
        Configuracao::instancia().exibir();
    }
};

// ── Main ─────────────────────────────────────────────────
int main() {
    MenuPrincipal menu;
    Masmorra      masmorra;

    menu.aplicar_configuracoes();
    masmorra.iniciar();   // ✅ enxerga "dificil" e volume 60

    auto& a = Configuracao::instancia();
    auto& b = Configuracao::instancia();
    std::cout << "\nMesma instância? "
              << (&a == &b ? "True ✓" : "False ✗") << "\n";
    return 0;
}

File: codes/semana11/exercicio02.py
from __future__ import annotations

# ── Singleton ────────────────────────────────────────────
class Configuracao:
    _instancia: Configuracao | None = None

    def __new__(cls) -> Configuracao:
        if cls._instancia is None:
            cls._instancia = super().__new__(cls)
            cls._instancia._inicializado = False
        return cls._instancia

    def __init__(self):
        if self._inicializado:
            return
        self._dados: dict[str, str] = {
            "dificuldade": "normal",
            "volume":      "80",
            "idioma":      "pt-BR",
        }
        self._inicializado = True

    @classmethod
    def instancia(cls) -> Configuracao:
        return cls()

    def alterar(self, chave: str, valor: str) -> None:
        self._dados[chave] = valor

    def obter(self, chave: str) -> str:
        return self._dados[chave]

    def exibir(self) -> None:
        for chave, valor in self._dados.items():
            print(f"  [Config] {chave:<12} = {valor}")

# ── Classes consumidoras ─────────────────────────────────
class MenuPrincipal:
    def aplicar_configuracoes(self) -> None:
        cfg = Configuracao.instancia()
        cfg.alterar("dificuldade", "dificil")
        cfg.alterar("volume",      "60")
        print("[Menu] Configurações alteradas pelo jogador")
        cfg.exibir()

class Masmorra:
    def iniciar(self) -> None:
        print("\n[Masmorra] Lendo configurações do jogo...")
        Configuracao.instancia().exibir()  # ✅ vê "dificil"

# ── Uso ──────────────────────────────────────────────────
menu     = MenuPrincipal()
masmorra = Masmorra()

menu.aplicar_configuracoes()
masmorra.iniciar()

# Prova: mesmo objeto
a = Configuracao()
b = Configuracao()
print(f"\nMesma instância? {a is b} ✓")